csrc/scheduler/normalization_utils.cpp

```diff
@@ -813,10 +813,43 @@ void checkReductionTvForScheduling(Fusion* fusion, TensorView* ref_red_tv) {
       "Tried to schedule a fusion with no tensor inputs, currently not supported.");
 }
 
+int64_t getMaxRegOrSharedMemorySizeForPersistentBuffer(
+    SchedulerRuntimeInfo& runtime_info,
+    const std::vector<TensorView*>& persistent_buffers) {
+  // Init to register file size, which is half of the full register file size
+  int64_t available_persistent_buffer_size =
+      scheduler_utils::register_file_size;
+
+  // Check available shared memory
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t max_shared_memory_size =
+      (int64_t)dev_prop->sharedMemPerBlockOptin;
+  // Some shared memories are reserved for kernel launch overhead and
+  // reduction_broadcast_workspace. Estimation is conservative, but should
+  // be good enough. The actual threads per block is set in the heuristics
+  // and it may be smaller than maxThreadsPerBlock.
+  // TODO: More accurate estimation of available shared memory size.
+  const int64_t kernel_overhead = (int64_t)dev_prop->reservedSharedMemPerBlock;
+  int64_t max_buffer_dtype_size = 1;
+  for (auto tv : persistent_buffers) {
+    max_buffer_dtype_size = std::max(
+        max_buffer_dtype_size,
+        dataTypeSize(tv->getDataType().value(), runtime_info.getIndexType()));
+  }
+  const int64_t reduction_broadcast_workspace =
+      (int64_t)(dev_prop->maxThreadsPerBlock) * max_buffer_dtype_size;
+  const int64_t available_shared_memory_size =
+      max_shared_memory_size - kernel_overhead - reduction_broadcast_workspace;
+  available_persistent_buffer_size =
+      std::max(available_persistent_buffer_size, available_shared_memory_size);
+  return available_persistent_buffer_size;
+}
+
 // Returns true if the gains of reducing buffer size is larger than the pains of
 // recalculations. We don't know the real answer until we run it.
```

Comment:
Could you please update this comment? This function becomes more complex since it was initially written.

Reply:
make sense. added in a separate PR. https://github.com/NVIDIA/Fuser/pull/1961